scripts/lib/source.py

```diff
@@ -0,0 +1,219 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import fnmatch
+import logging
+import os
+import re
+
+from lib import trace
+
+
+# Paths under which all files should be ignored
+IGNORE = frozenset([
+    'Firestore/Protos/nanopb',
+    'Firestore/Protos/cpp',
+    'Firestore/Protos/objc',
+    'Firestore/third_party/abseil-cpp',
+    'GoogleDataTransportCCTSupport/GDTCCTLibrary/Protogen/nanopb',
+    'GoogleDataTransportCCTSupport/ProtoSupport',
+])
+
+FIRESTORE_CORE = ['Firestore/core']
+FIRESTORE_OBJC = ['Firestore/Source', 'Firestore/Example/Tests']
+FIRESTORE_SWIFT = ['Firestore/Swift']
+
+FIRESTORE_TESTS = ['Firestore/core/test', 'Firestore/Example/Tests']
+
+CC_DIRS = FIRESTORE_CORE
+CC_EXTENSIONS = ['.h', '.cc']
+
+OBJC_DIRS = FIRESTORE_CORE + FIRESTORE_OBJC
+OBJC_EXTENSIONS = ['.h', '.m', '.mm']
+
+PYTHON_DIRS = ['scripts']
+PYTHON_EXTENSIONS = ['.py']
+
+SOURCE_EXTENSIONS = [
+    '.c',
+    '.cc',
+    '.cmake',
+    '.h',
+    '.js',
+    '.m',
+    '.mm',
+    '.py',
+    '.rb',
+    '.sh',
+    '.swift'
+]
+
+_DEFINITE_EXTENSIONS = {
+    '.cc': 'cc',
+    '.m': 'objc',
+    '.mm': 'objc',
+    '.py': 'py',
+}
+
+
+_classify_logger = logging.getLogger('lint.classify')
+
+
+class LanguageBreakdown:
+  """Files broken down by source language."""
+
+  def __init__(self):
+    self.cc = []
+    self.objc = []
+    self.py = []
+    self.all = []
+
+    self.kinds = {
+        'cc': self.cc,
+        'objc': self.objc,
+        'py': self.py,
+    }
+
+  def classify(self, kind, reason, filename):
+    _classify_logger.debug('classify %s: %s (%s)' % (kind, filename, reason))
+    self.kinds[kind].append(filename)
+    self.all.append(filename)
+
+  @staticmethod
+  def ignore(filename):
+    _classify_logger.debug('classify ignored: %s' % filename)
+
+
+def categorize_files(files):
+  """Breaks down the given list of files by language.
+
+  Args:
+    files: a list of files
+
+  Returns:
+    A LanguageBreakdown instance containing all the files that match a
+    recognized source language.
+  """
+  result = LanguageBreakdown()
+
+  for filename in files:
+    if _in_directories(filename, IGNORE):
+      continue
+
+    ext = os.path.splitext(filename)[1]
+    definite = _DEFINITE_EXTENSIONS.get(ext)
+    if definite:
+      result.classify(definite, 'extension', filename)
+      continue
+
+    if ext == '.h':
+      if _in_directories(filename, CC_DIRS):
+        result.classify('cc', 'directory', filename)
+        continue
+
+      related_ext = _related_file_ext(filename)
+      if related_ext == '.cc':
+        result.classify('cc', 'related file', filename)
+        continue
+
+      if related_ext in ('.m', '.mm'):
+        result.classify('objc', 'related file', filename)
+        continue
```

Comment:
If we have a header-only file (outside of Firestore/core), I think this will cause it to not be linted?